RoddyCore/src/de/dkfz/roddy/client/cliclient/CommandLineCall.groovy

```diff
@@ -6,135 +6,248 @@
-import de.dkfz.roddy.execution.io.ExecutionService
 import de.dkfz.roddy.tools.LoggerWrapper
-
-import static de.dkfz.roddy.StringConstants.SPLIT_COMMA
-import static de.dkfz.roddy.client.RoddyStartupModes.help
+import de.dkfz.roddy.tools.Tuple2
+import groovy.transform.CompileDynamic
+import org.petitparser.context.Result
+import org.petitparser.parser.Parser
 
 /**
  * A helper class for command line calls.
  * Options and parameters are extracted and put into an object of this class.
  */
 @groovy.transform.CompileStatic
-public class CommandLineCall {
-    private static LoggerWrapper logger = LoggerWrapper.getLogger(RoddyCLIClient.class.getSimpleName());
-    public final RoddyStartupModes startupMode;
-    private final List<String> arguments;
-    private final List<String> parameters;
-    private Map<RoddyStartupOptions, List<String>> optionsMap;
+class CommandLineCall {
+
+    private static final Parser cvaluesParser = CommandLineParameterParser.cvalueParameterExpr.end("malformed content")
+    private static final Parser otherParser = CommandLineParameterParser.parameterWithValueExpr.
+            or(CommandLineParameterParser.parameterWithoutValueExpr).
+            or(CommandLineParameterParser.arbitraryParameterExpr).end()
+
+
+    private static LoggerWrapper logger = LoggerWrapper.getLogger(RoddyCLIClient.class.getSimpleName())
+    public final RoddyStartupModes startupMode
+    private final List<String> arguments
+    private final List<String> parameters
+    private Map<RoddyStartupOptions, Parameter> optionsMap
     final boolean malformed
 
-    CommandLineCall(List<String> args) {
-        parameters = [];
-        optionsMap = [:];
-        if (!args.size() == 0)
-            args = [RoddyStartupModes.help.toString()];
-        this.arguments = args;
-
-        List<String> parameters = args.findAll { String arg -> !arg.startsWith("--") } as ArrayList<String>;
-        Collection<String> options = args.findAll { String arg -> arg.startsWith("--") };
-
-        // Try to extract the startup mode. If it not know, display the help message.
-        try {
-            startupMode = parameters ? parameters[0] as RoddyStartupModes : help;
-        } catch (Exception ex) {
-            logger.postAlwaysInfo("The startupmode " + parameters[0] + " is not known.");
-            this.startupMode = help;
-            return;
+    static RoddyStartupModes parseStartupMode(List<String> args) {
+        // The following implies that parameters (without '--') and options (with '--') can be freely mixed.
+        String putativeModeSpec = args.find { String arg -> !arg.startsWith('--') }
+        Optional<RoddyStartupModes> mode = RoddyStartupModes.fromString(putativeModeSpec)
+        if (mode.isPresent()) {
+            return mode.get()
+        } else {
+            logger.postAlwaysInfo("The startupmode '${putativeModeSpec}' is not known.")
+            return help
         }
+    }
 
-        //Parse options
+    @CompileDynamic
```

Comment:
This is not actually really part of the parser stuff. I used the @CompileDynamic to not have to use ```instanceof``` in switch or if-else. The code in these functions glues together the RoddyStartupOption code with the Parameter-subclasses from the parser.
Until now I did not consider to extract them into another class but stayed close to the orginal code. I will think about it.

Reply:
Hm, maybe instanceof would be simpler as the current code as quite some boilerplate code. Otherwise I do not see room for substantial improvement, and a bit of boilerplate-code also is not that big of an issue and has just other advantages and disadvantages compared to instanceof. Maybe there are more important points to improve in the code. Otherwise I suggest we leave it like this and maybe change the code when we switch to a actual command-line parser (outside) with the current cvalues-Parameter parser nested inside only for the cvalues-Parameter.